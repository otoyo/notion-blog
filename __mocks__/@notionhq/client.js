'use strict'

import * as fs from 'fs'
import * as path from 'path'

const client = jest.createMockFromModule('@notionhq/client')

client.Client = class {
  constructor() {
    return {
      databases: {
        query: () => {
          return JSON.parse(
            fs
              .readFileSync(
                path.resolve(
                  './__tests__/fixtures/notion-api-response-pages.json'
                )
              )
              .toString()
          )
        },
        retrieve: () => {
          return JSON.parse(
            fs
              .readFileSync(
                path.resolve(
                  './__tests__/fixtures/notion-api-response-database.json'
                )
              )
              .toString()
          )
        },
      },
      blocks: {
        children: {
          list: ({ block_id }) => {
            const tableBlockId = '30f2f940-3bd8-46de-b76c-00e4c0eb9521'

            if (block_id === tableBlockId) {
              return JSON.parse(
                fs
                  .readFileSync(
                    path.resolve(
                      './__tests__/fixtures/notion-api-response-table-row-blocks.json'
                    )
                  )
                  .toString()
              )
            }
            return JSON.parse(
              fs
                .readFileSync(
                  path.resolve(
                    './__tests__/fixtures/notion-api-response-page-blocks.json'
                  )
                )
                .toString()
            )
          },
        },
      },
    }
  }
}

module.exports = client
