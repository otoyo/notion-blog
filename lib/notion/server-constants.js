/* eslint @typescript-eslint/no-var-requires: 0 */
// use commonjs so it can be required without transpiling

const NOTION_API_SECRET = process.env.NOTION_API_SECRET
const DATABASE_ID = process.env.DATABASE_ID
const INDEX_PAGE_ID = process.env.INDEX_PAGE_ID
const SUBSCRIPTION_PAGE_ID = process.env.SUBSCRIPTION_PAGE_ID
const NEXT_PUBLIC_URL = process.env.NEXT_PUBLIC_URL
const NEXT_PUBLIC_GA_TRACKING_ID = process.env.NEXT_PUBLIC_GA_TRACKING_ID
const NEXT_PUBLIC_SITE_TITLE = 'アルパカログ'
const NEXT_PUBLIC_SITE_DESCRIPTION = 'Notion Blogのカスタマイズ、マネジメント、プログラミングや読んだ本のまとめなどが中心のブログ'
const NUMBER_OF_POSTS_PER_PAGE = 10

module.exports = {
  NOTION_API_SECRET,
  DATABASE_ID,
  INDEX_PAGE_ID,
  SUBSCRIPTION_PAGE_ID,
  NEXT_PUBLIC_URL,
  NEXT_PUBLIC_GA_TRACKING_ID,
  NEXT_PUBLIC_SITE_TITLE,
  NEXT_PUBLIC_SITE_DESCRIPTION,
  NUMBER_OF_POSTS_PER_PAGE,
}
