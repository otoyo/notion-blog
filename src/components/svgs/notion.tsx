export default props => (
  <svg
    viewBox="-5 0 120 130"
    className="prefix__notionLogo"
    display="block"
    fill="inherit"
    {...props}
  >
    <path d="M20.693 21.931c3.89 3.161 5.35 2.92 12.656 2.433l68.879-4.136c1.461 0 .246-1.458-.241-1.7l-11.44-8.27c-2.191-1.701-5.111-3.65-10.708-3.162L13.143 11.96c-2.432.241-2.918 1.457-1.95 2.432l9.5 7.54zm4.135 16.052v72.473c0 3.895 1.947 5.352 6.327 5.111l75.698-4.38c4.383-.241 4.871-2.92 4.871-6.084V33.117c0-3.159-1.215-4.863-3.898-4.62l-79.105 4.62c-2.92.245-3.893 1.706-3.893 4.867zm74.729 3.888c.485 2.191 0 4.38-2.195 4.626l-3.648.727v53.504c-3.166 1.702-6.087 2.675-8.52 2.675-3.896 0-4.871-1.217-7.79-4.863L53.547 61.087v36.237l7.55 1.704s0 4.375-6.091 4.375l-16.791.974c-.488-.974 0-3.404 1.703-3.891l4.382-1.214V51.36l-6.084-.487c-.488-2.191.727-5.35 4.137-5.596l18.013-1.214 24.828 37.94V48.44l-6.33-.726c-.486-2.679 1.459-4.624 3.893-4.865l16.8-.978zM7.543 5.394L76.917.285c8.52-.73 10.712-.241 16.066 3.649l22.145 15.564c3.654 2.677 4.872 3.405 4.872 6.323v85.366c0 5.35-1.949 8.514-8.763 8.998l-80.564 4.865c-5.115.244-7.549-.485-10.228-3.892L4.137 99.999C1.215 96.105 0 93.191 0 89.782V13.903c0-4.375 1.95-8.025 7.543-8.509z" />
  </svg>
)
