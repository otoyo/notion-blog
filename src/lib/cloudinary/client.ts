import cloudinary from 'cloudinary'
import {
  CLOUDINARY_CLOUD_NAME,
  CLOUDINARY_API_KEY,
  CLOUDINARY_API_SECRET,
} from '../../server-constants'

function existsConfig(): boolean {
  return !!(CLOUDINARY_CLOUD_NAME && CLOUDINARY_API_KEY && CLOUDINARY_API_SECRET)
}

if (existsConfig()) {
  cloudinary.v2.config({
    cloud_name: CLOUDINARY_CLOUD_NAME,
    api_key: CLOUDINARY_API_KEY,
    api_secret: CLOUDINARY_API_SECRET,
  })
}


export async function uploadImageWithKey(url: string, key: string): Promise<string> {
  if (!existsConfig() || !url || !key) {
    return ''
  }

  try {
    const res = await cloudinary.v2.uploader.upload(url, { public_id: key })
    return res.secure_url
  } catch (err) {
    console.log(err)
    return ''
  }
}

export function imageURLWithKey(key: string): string {
  if (!existsConfig() || !key) {
    return ''
  }

  return cloudinary.v2.url(key)
}
