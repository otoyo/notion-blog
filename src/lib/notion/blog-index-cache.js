const path = require('path')
const fs = require('fs')

const {
  NOTION_API_SECRET,
  DATABASE_ID,
  BLOG_INDEX_CACHE,
} = require('./server-constants')
const { Client } = require('@notionhq/client')

const notionClient = new Client({
  auth: NOTION_API_SECRET,
})

exports.exists = function() {
  return fs.existsSync(path.resolve(BLOG_INDEX_CACHE))
}

exports.get = function() {
  return JSON.parse(fs.readFileSync(path.resolve(BLOG_INDEX_CACHE)))
}

exports.set = async function() {
  let params = {
    database_id: DATABASE_ID,
    filter: {
      and: [
        {
          property: 'Published',
          checkbox: {
            equals: true,
          },
        },
        {
          property: 'Date',
          date: {
            on_or_before: new Date().toISOString(),
          },
        },
      ],
    },
    sorts: [
      {
        property: 'Date',
        timestamp: 'created_time',
        direction: 'descending',
      },
    ],
    page_size: 100,
  }

  let results = []

  while (true) {
    const data = await notionClient.databases.query(params)

    results = results.concat(data.results)

    if (!data.has_more) {
      break
    }

    params['start_cursor'] = data.next_cursor
  }

  fs.writeFileSync(path.resolve(BLOG_INDEX_CACHE), JSON.stringify(results))
  console.log(`Cached ${results.length} posts into ${BLOG_INDEX_CACHE}`)

  return
}
