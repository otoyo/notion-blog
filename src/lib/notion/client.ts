import fs from 'node:fs'
import {
  NOTION_API_SECRET,
  DATABASE_ID,
  NUMBER_OF_POSTS_PER_PAGE,
} from '../../server-constants'
import type * as responses from './responses'
import type * as requestParams from './request-params'
import type {
  Post,
  Block,
  Paragraph,
  Heading1,
  Heading2,
  Heading3,
  BulletedListItem,
  NumberedListItem,
  ToDo,
  Image,
  Code,
  Quote,
  Equation,
  Callout,
  Embed,
  Video,
  Bookmark,
  LinkPreview,
  SyncedBlock,
  SyncedFrom,
  Table,
  TableRow,
  TableCell,
  Toggle,
  ColumnList,
  Column,
  TableOfContents,
  RichText,
  Text,
  Annotation,
  SelectProperty,
} from '../interfaces'
// eslint-disable-next-line @typescript-eslint/no-var-requires
import { Client } from '@notionhq/client'

const client = new Client({
  auth: NOTION_API_SECRET,
})

let cache: Post[] | null = null


export async function getAllPosts(): Promise<Post[]> {
  if (cache !== null) {
    return Promise.resolve(cache)
  }

  const params: requestParams.QueryDatabase = {
    database_id: DATABASE_ID,
    filter: {
      and: [
        {
          property: 'Published',
          checkbox: {
            equals: true,
          },
        },
        {
          property: 'Date',
          date: {
            on_or_before: new Date().toISOString(),
          },
        },
      ],
    },
    sorts: [
      {
        property: 'Date',
        direction: 'descending',
      },
    ],
    page_size: 100,
  }

  let results: responses.PageObject[] = []
  while (true) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const res = await client.databases.query(params as any) as responses.QueryDatabaseResponse

    results = results.concat(res.results)

    if (!res.has_more) {
      break
    }

    params['start_cursor'] = res.next_cursor as string
  }

  cache = results
    .filter(pageObject => _validPageObject(pageObject))
    .map(pageObject => _buildPost(pageObject))
  return cache
}

export async function getPosts(pageSize = 10): Promise<Post[]> {
  const allPosts = await getAllPosts()
  return allPosts.slice(0, pageSize)
}

export async function getRankedPosts(pageSize = 10): Promise<Post[]> {
  const allPosts = await getAllPosts()
  return allPosts
    .filter(post => !!post.Rank)
    .sort((a, b) => {
      if (a.Rank > b.Rank) {
        return -1
      } else if (a.Rank === b.Rank) {
        return 0
      }
      return 1
    })
    .slice(0, pageSize)
}

export async function getPostBySlug(slug: string): Promise<Post|null> {
  const allPosts = await getAllPosts()
  return allPosts.find(post => post.Slug === slug) || null
}

export async function getPostsByTag(tagName: string, pageSize = 10): Promise<Post[]> {
  if (!tagName) return []

  const allPosts = await getAllPosts()
  return allPosts.filter(post => post.Tags.find((tag) => tag.name === tagName)).slice(0, pageSize)
}

// page starts from 1 not 0
export async function getPostsByPage(page: number): Promise<Post[]> {
  if (page < 1) {
    return []
  }

  const allPosts = await getAllPosts()

  const startIndex = (page - 1) * NUMBER_OF_POSTS_PER_PAGE
  const endIndex = startIndex + NUMBER_OF_POSTS_PER_PAGE

  return allPosts.slice(startIndex, endIndex)
}

// page starts from 1 not 0
export async function getPostsByTagAndPage(tagName: string, page: number): Promise<Post[]> {
  if (page < 1) {
    return []
  }

  const allPosts = await getAllPosts()
  const posts = allPosts.filter(post => post.Tags.find((tag) => tag.name === tagName))

  const startIndex = (page - 1) * NUMBER_OF_POSTS_PER_PAGE
  const endIndex = startIndex + NUMBER_OF_POSTS_PER_PAGE

  return posts.slice(startIndex, endIndex)
}

export async function getNumberOfPages(): Promise<number> {
  const allPosts = await getAllPosts()
  return Math.floor(allPosts.length / NUMBER_OF_POSTS_PER_PAGE) + (allPosts.length % NUMBER_OF_POSTS_PER_PAGE > 0 ? 1 : 0)
}

export async function getNumberOfPagesByTag(tagName: string): Promise<number> {
  const allPosts = await getAllPosts()
  const posts = allPosts.filter(post => post.Tags.find((tag) => tag.name === tagName))
  return Math.floor(posts.length / NUMBER_OF_POSTS_PER_PAGE) + (posts.length % NUMBER_OF_POSTS_PER_PAGE > 0 ? 1 : 0)
}

export async function getAllBlocksByBlockId(blockId: string): Promise<Block[]> {
  let results: responses.BlockObject[] = []

  if (fs.existsSync(`tmp/${blockId}.json`)) {
    results = JSON.parse(fs.readFileSync(`tmp/${blockId}.json`, 'utf-8'))
  } else {
    const params: requestParams.RetrieveBlockChildren = {
      block_id: blockId,
    }

    while (true) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const res = await client.blocks.children.list(params as any) as responses.RetrieveBlockChildrenResponse

      results = results.concat(res.results)

      if (!res.has_more) {
        break
      }

      params['start_cursor'] = res.next_cursor as string
    }
  }

  const allBlocks = results.map(blockObject => _buildBlock(blockObject))

  for (let i = 0; i < allBlocks.length; i++) {
    const block = allBlocks[i]

    if (block.Type === 'table' && block.Table) {
      block.Table.Rows = await _getTableRows(block.Id)
    } else if (block.Type === 'column_list' && block.ColumnList) {
      block.ColumnList.Columns = await _getColumns(block.Id)
    } else if (block.Type === 'bulleted_list_item' && block.BulletedListItem && block.HasChildren) {
      block.BulletedListItem.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'numbered_list_item' && block.NumberedListItem && block.HasChildren) {
      block.NumberedListItem.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'to_do' && block.ToDo && block.HasChildren) {
      block.ToDo.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'synced_block' && block.SyncedBlock) {
      block.SyncedBlock.Children = await _getSyncedBlockChildren(block)
    } else if (block.Type === 'toggle' && block.Toggle) {
      block.Toggle.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'paragraph' && block.Paragraph && block.HasChildren) {
      block.Paragraph.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'heading_1' && block.Heading1 && block.HasChildren) {
      block.Heading1.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'heading_2' && block.Heading2 && block.HasChildren) {
      block.Heading2.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'heading_3' && block.Heading3 && block.HasChildren) {
      block.Heading3.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'quote' && block.Quote && block.HasChildren) {
      block.Quote.Children = await getAllBlocksByBlockId(block.Id)
    } else if (block.Type === 'callout' && block.Callout && block.HasChildren) {
      block.Callout.Children = await getAllBlocksByBlockId(block.Id)
    }
  }

  return allBlocks
}

export async function getBlock(blockId: string): Promise<Block> {
  const params: requestParams.RetrieveBlock = {
    block_id: blockId,
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const res = await client.blocks.retrieve(params as any) as responses.RetrieveBlockResponse

  return _buildBlock(res)
}

export async function getAllTags(): Promise<SelectProperty[]> {
  const allPosts = await getAllPosts()

  const tagNames: string[] = []
  return allPosts.flatMap(post => post.Tags).reduce((acc, tag) => {
    if (!tagNames.includes(tag.name)) {
      acc.push(tag)
      tagNames.push(tag.name)
    }
    return acc
  }, [] as SelectProperty[]).sort((a: SelectProperty, b: SelectProperty) => a.name.localeCompare(b.name))
}

function _buildBlock(blockObject: responses.BlockObject): Block {
  const block: Block = {
    Id: blockObject.id,
    Type: blockObject.type,
    HasChildren: blockObject.has_children,
  }

  switch (blockObject.type) {
    case 'paragraph':
      if (blockObject.paragraph) {
        const paragraph: Paragraph = {
          RichTexts: blockObject.paragraph.rich_text.map(_buildRichText),
          Color: blockObject.paragraph.color,
        }
        block.Paragraph = paragraph
      }
      break
    case 'heading_1':
      if (blockObject.heading_1) {
        const heading1: Heading1 = {
          RichTexts: blockObject.heading_1.rich_text.map(_buildRichText),
          Color: blockObject.heading_1.color,
          IsToggleable: blockObject.heading_1.is_toggleable,
        }
        block.Heading1 = heading1
      }
      break
    case 'heading_2':
      if (blockObject.heading_2) {
        const heading2: Heading2 = {
          RichTexts: blockObject.heading_2.rich_text.map(_buildRichText),
          Color: blockObject.heading_2.color,
          IsToggleable: blockObject.heading_2.is_toggleable,
        }
        block.Heading2 = heading2
      }
      break
    case 'heading_3':
      if (blockObject.heading_3) {
        const heading3: Heading3 = {
          RichTexts: blockObject.heading_3.rich_text.map(_buildRichText),
          Color: blockObject.heading_3.color,
          IsToggleable: blockObject.heading_3.is_toggleable,
        }
        block.Heading3 = heading3
      }
      break
    case 'bulleted_list_item':
      if (blockObject.bulleted_list_item) {
        const bulletedListItem: BulletedListItem = {
          RichTexts: blockObject.bulleted_list_item.rich_text.map(_buildRichText),
          Color: blockObject.bulleted_list_item.color,
        }
        block.BulletedListItem = bulletedListItem
      }
      break
    case 'numbered_list_item':
      if (blockObject.numbered_list_item) {
        const numberedListItem: NumberedListItem = {
          RichTexts: blockObject.numbered_list_item.rich_text.map(_buildRichText),
          Color: blockObject.numbered_list_item.color,
        }
        block.NumberedListItem = numberedListItem
      }
      break
    case 'to_do':
      if (blockObject.to_do) {
        const toDo: ToDo = {
          RichTexts: blockObject.to_do.rich_text.map(_buildRichText),
          Checked: blockObject.to_do.checked,
          Color: blockObject.to_do.color,
        }
        block.ToDo = toDo
      }
      break
    case 'video':
      if (blockObject.video) {
        const video: Video = {
          Caption: blockObject.video.caption?.map(_buildRichText) || [],
          Type: blockObject.video.type,
        }
        if (blockObject.video.type === 'external' && blockObject.video.external) {
          video.External = { Url: blockObject.video.external.url }
        }
        block.Video = video
      }
      break
    case 'image':
      if (blockObject.image) {
        const image: Image = {
          Caption: blockObject.image.caption?.map(_buildRichText) || [],
          Type: blockObject.image.type,
        }
        if (blockObject.image.type === 'external' && blockObject.image.external) {
          image.External = { Url: blockObject.image.external.url }
        } else if (blockObject.image.type === 'file' && blockObject.image.file) {
          image.File = { Url: blockObject.image.file.url, ExpiryTime: blockObject.image.file.expiry_time }
        }
        block.Image = image
      }
      break
    case 'code':
      if (blockObject.code) {
        const code: Code = {
          Caption: blockObject.code.caption?.map(_buildRichText) || [],
          RichTexts: blockObject.code.rich_text.map(_buildRichText),
          Language: blockObject.code.language,
        }
        block.Code = code
      }
      break
    case 'quote':
      if (blockObject.quote) {
        const quote: Quote = {
          RichTexts: blockObject.quote.rich_text.map(_buildRichText),
          Color: blockObject.quote.color,
        }
        block.Quote = quote
      }
      break
    case 'equation':
      if (blockObject.equation) {
        const equation: Equation = {
          Expression: blockObject.equation.expression,
        }
        block.Equation = equation
      }
      break
    case 'callout':
      if (blockObject.callout) {
        const callout: Callout = {
          RichTexts: blockObject.callout.rich_text.map(_buildRichText),
          Icon: {
            Emoji: blockObject.callout.icon.emoji,
          },
          Color: blockObject.callout.color,
        }
        block.Callout = callout
      }
      break
    case 'synced_block':
      if (blockObject.synced_block) {
        let syncedFrom: SyncedFrom | null = null
        if (blockObject.synced_block.synced_from && blockObject.synced_block.synced_from.block_id) {
          syncedFrom = {
            BlockId: blockObject.synced_block.synced_from.block_id,
          }
        }

        const syncedBlock: SyncedBlock = {
          SyncedFrom: syncedFrom,
        }
        block.SyncedBlock = syncedBlock
      }
      break
    case 'toggle':
      if (blockObject.toggle) {
        const toggle: Toggle = {
          RichTexts: blockObject.toggle.rich_text.map(_buildRichText),
          Color: blockObject.toggle.color,
          Children: [],
        }
        block.Toggle = toggle
      }
      break
    case 'embed':
      if (blockObject.embed) {
        const embed: Embed = {
          Url: blockObject.embed.url,
        }
        block.Embed = embed
      }
      break
    case 'bookmark':
      if (blockObject.bookmark) {
        const bookmark: Bookmark = {
          Url: blockObject.bookmark.url,
        }
        block.Bookmark = bookmark
      }
      break
    case 'link_preview':
      if (blockObject.link_preview) {
        const linkPreview: LinkPreview = {
          Url: blockObject.link_preview.url,
        }
        block.LinkPreview = linkPreview
      }
      break
    case 'table':
      if (blockObject.table) {
        const table: Table = {
          TableWidth: blockObject.table.table_width,
          HasColumnHeader: blockObject.table.has_column_header,
          HasRowHeader: blockObject.table.has_row_header,
          Rows: [],
        }
        block.Table = table
      }
      break
    case 'column_list':
      const columnList: ColumnList = {
        Columns: [],
      }
      block.ColumnList = columnList
      break
    case 'table_of_contents':
      if (blockObject.table_of_contents) {
        const tableOfContents: TableOfContents = {
          Color: blockObject.table_of_contents.color,
        }
        block.TableOfContents = tableOfContents
      }
      break
  }

  return block
}

async function _getTableRows(blockId: string): Promise<TableRow[]> {
  let results: responses.BlockObject[] = []

  if (fs.existsSync(`tmp/${blockId}.json`)) {
    results = JSON.parse(fs.readFileSync(`tmp/${blockId}.json`, 'utf-8'))
  } else {
    const params: requestParams.RetrieveBlockChildren = {
      block_id: blockId,
    }

    while (true) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const res = await client.blocks.children.list(params as any) as responses.RetrieveBlockChildrenResponse

      results = results.concat(res.results)

      if (!res.has_more) {
        break
      }

      params['start_cursor'] = res.next_cursor as string
    }
  }

  return results.map(blockObject => {
    const tableRow: TableRow = {
      Id: blockObject.id,
      Type: blockObject.type,
      HasChildren: blockObject.has_children,
      Cells: []
    }

    if (blockObject.type === 'table_row' && blockObject.table_row) {
      const cells: TableCell[] = blockObject.table_row.cells.map(cell => {
        const tableCell: TableCell = {
          RichTexts: cell.map(_buildRichText),
        }

        return tableCell
      })

      tableRow.Cells = cells
    }

    return tableRow
  })
}

async function _getColumns(blockId: string): Promise<Column[]> {
  let results: responses.BlockObject[] = []

  if (fs.existsSync(`tmp/${blockId}.json`)) {
    results = JSON.parse(fs.readFileSync(`tmp/${blockId}.json`, 'utf-8'))
  } else {
    const params: requestParams.RetrieveBlockChildren = {
      block_id: blockId,
    }

    while (true) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const res = await client.blocks.children.list(params as any) as responses.RetrieveBlockChildrenResponse

      results = results.concat(res.results)

      if (!res.has_more) {
        break
      }

      params['start_cursor'] = res.next_cursor as string
    }
  }

  return await Promise.all(results.map(async blockObject => {
    const children = await getAllBlocksByBlockId(blockObject.id)

    const column: Column = {
      Id: blockObject.id,
      Type: blockObject.type,
      HasChildren: blockObject.has_children,
      Children: children,
    }

    return column
  }))
}

async function _getSyncedBlockChildren(block: Block): Promise<Block[]> {
  let originalBlock: Block = block
  if (block.SyncedBlock && block.SyncedBlock.SyncedFrom && block.SyncedBlock.SyncedFrom.BlockId) {
    try {
      originalBlock = await getBlock(block.SyncedBlock.SyncedFrom.BlockId)
    } catch (err) {
      console.log(`Could not retrieve the original synced_block. error: ${err}`)
      return []
    }
  }

  const children = await getAllBlocksByBlockId(originalBlock.Id)
  return children
}

function _validPageObject(pageObject: responses.PageObject): boolean {
  const prop = pageObject.properties
  return (
    !!prop.Page.title && prop.Page.title.length > 0 &&
    !!prop.Slug.rich_text && prop.Slug.rich_text.length > 0 &&
    !!prop.Date.date
  )
}

function _buildPost(pageObject: responses.PageObject): Post {
  const prop = pageObject.properties

  const post: Post = {
    PageId: pageObject.id,
    Title: prop.Page.title ? prop.Page.title[0].plain_text : '',
    Slug: prop.Slug.rich_text ? prop.Slug.rich_text[0].plain_text : '',
    Date: prop.Date.date ? prop.Date.date.start : '',
    Tags: prop.Tags.multi_select ? prop.Tags.multi_select : [],
    Excerpt:
      prop.Excerpt.rich_text && prop.Excerpt.rich_text.length > 0
        ? prop.Excerpt.rich_text.map(t => t.plain_text).join('')
        : '',
    FeaturedImage:
      prop.FeaturedImage.files && prop.FeaturedImage.files.length > 0 && prop.FeaturedImage.files[0].file
        ? prop.FeaturedImage.files[0].file.url
        : null,
    Rank: prop.Rank.number ? prop.Rank.number : 0,
  }

  return post
}

function _buildRichText(richTextObject: responses.RichTextObject): RichText {
  const annotation: Annotation = {
    Bold: richTextObject.annotations.bold,
    Italic: richTextObject.annotations.italic,
    Strikethrough: richTextObject.annotations.strikethrough,
    Underline: richTextObject.annotations.underline,
    Code: richTextObject.annotations.code,
    Color: richTextObject.annotations.color,
  }

  const richText: RichText = {
    Annotation: annotation,
    PlainText: richTextObject.plain_text,
    Href: richTextObject.href,
  }

  if (richTextObject.type === 'text' && richTextObject.text) {
    const text: Text = {
      Content: richTextObject.text.content,
    }

    if (richTextObject.text.link) {
      text.Link = {
        Url: richTextObject.text.link.url,
      }
    }

    richText.Text = text
  } else if (richTextObject.type === 'equation' && richTextObject.equation) {
    const equation: Equation = {
      Expression: richTextObject.equation.expression,
    }
    richText.Equation = equation
  }

  return richText
}
