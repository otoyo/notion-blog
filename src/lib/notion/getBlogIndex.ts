import { Sema } from 'async-sema'
import rpc, { values } from './rpc'
import createTable from './createTable'
import getTableData from './getTableData'
import { getPostPreview } from './getPostPreview'
import { readFile, writeFile } from '../fs-helpers'
import { BLOG_INDEX_ID, BLOG_INDEX_CACHE } from './server-constants'

export default async function getBlogIndex(previews = true) {
  let postsTable: any = null
  const useCache = process.env.USE_CACHE === 'true'
  const cacheFile = `${BLOG_INDEX_CACHE}${previews ? '_previews' : ''}`

  if (useCache) {
    try {
      postsTable = JSON.parse(await readFile(cacheFile, 'utf8'))
    } catch (_) {
      /* not fatal */
    }
  }

  if (!postsTable) {
    try {
      const data = await rpc('loadPageChunk', {
        pageId: BLOG_INDEX_ID,
        limit: 999, // TODO: figure out Notion's way of handling pagination
        cursor: { stack: [] },
        chunkNumber: 0,
        verticalColumns: false,
      })

      // Parse table with posts
      const tableBlock = values(data.recordMap.block).find(
        (block: any) => block.value.type === 'collection_view'
      )

      postsTable = await getTableData(tableBlock, true)
    } catch (err) {
      console.warn(
        `Failed to load Notion posts, attempting to auto create table`
      )
      try {
        await createTable()
        console.log(`Successfully created table in Notion`)
      } catch (err) {
        console.error(
          `Auto creating table failed, make sure you created a blank page and site the id with BLOG_INDEX_ID in your environment`,
          err
        )
      }
      return {}
    }

    // only get 10 most recent post's previews
    const postsKeys = Object.keys(postsTable).splice(0, 10)

    const sema = new Sema(3, { capacity: postsKeys.length })

    if (previews) {
      await Promise.all(
        postsKeys
          .sort((a, b) => {
            const postA = postsTable[a]
            const postB = postsTable[b]
            const timeA = postA.Date
            const timeB = postB.Date
            return Math.sign(timeB - timeA)
          })
          .map(async postKey => {
            await sema.acquire()
            const post = postsTable[postKey]
            post.preview = post.id
              ? await getPostPreview(postsTable[postKey].id)
              : []
            sema.release()
          })
      )
    }

    if (useCache) {
      writeFile(cacheFile, JSON.stringify(postsTable), 'utf8').catch(() => {})
    }
  }

  return postsTable
}
