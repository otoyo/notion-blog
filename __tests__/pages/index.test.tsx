import { render, waitFor } from '@testing-library/react'
import RenderPage from '../../src/pages/index'

import { getAllBlocksByBlockId } from '../../src/lib/notion/client'

jest.mock('next/router', () => ({
  useRouter() {
    return {
      asPath: '/',
      pathname: '/',
    }
  },
}))

describe('RenderPage', () => {
  it('renders the page unchanged', async () => {
    const blocks = await getAllBlocksByBlockId("")
    const { container } = render(<RenderPage blocks={blocks} />)
    await waitFor(() => {
      expect(container).toMatchSnapshot()
    })
  })
})
