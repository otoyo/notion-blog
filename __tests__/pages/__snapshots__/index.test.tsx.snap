// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderPage renders the page unchanged 1`] = `
<div>
  <div
    class="container"
  >
    <a
      class="heading"
      href="#Heading 1"
      id="Heading 1"
    >
      <h4>
        Heading 1
      </h4>
    </a>
    <a
      class="heading"
      href="#Heading 2"
      id="Heading 2"
    >
      <h5>
        Heading 2
      </h5>
    </a>
    <a
      class="heading"
      href="#Heading 3"
      id="Heading 3"
    >
      <h6>
        Heading 3
      </h6>
    </a>
    <p>
      Paragraph
    </p>
    <p>
      <b>
        bold
      </b>
    </p>
    <p>
      <i>
        italic
      </i>
    </p>
    <p>
      <u>
        Underline
      </u>
    </p>
    <p>
      <s>
        Strike
      </s>
    </p>
    <p>
      <a
        href="https://github.com/otoyo/easy-notion-blog"
      >
        link
      </a>
    </p>
    <p>
      <code>
        code
      </code>
    </p>
    <p />
    <ul>
      <li>
        Bulleted List
      </li>
      <li>
        Uni
      </li>
      <li>
        Tako
        <ul>
          <li>
            aaa
          </li>
          <li>
            bbb
          </li>
        </ul>
      </li>
    </ul>
    <p />
    <ol>
      <li>
        Numbered List
      </li>
      <li>
        Kani
      </li>
      <li>
        Ika
        <ol
          type="a"
        >
          <li>
            111
          </li>
          <li>
            222
          </li>
        </ol>
      </li>
    </ol>
    <p />
    <p>
      Divider
    </p>
    <hr
      class="divider"
    />
    <p />
    <div
      class="callout"
    >
      <div>
        ℹ️
      </div>
      <div>
        Callout
      </div>
    </div>
    <p />
    <p />
    <blockquote>
      Quote Block
    </blockquote>
    <p />
    <div
      class="table"
    >
      <table>
        <tbody>
          <tr>
            <th>
              Simple
            </th>
            <th>
              Table
            </th>
          </tr>
          <tr>
            <td>
              1
            </td>
            <td>
              Tako
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              Ika
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              Uni
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <p />
    <figure
      class="image"
    >
      <div>
        <img
          alt="画像が読み込まれない場合はページを更新してみてください。"
          src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/5d04be89-54da-482a-9391-efd465f74082/profile.png"
        />
      </div>
      <figcaption
        class="caption"
      >
        This is image caption.
      </figcaption>
    </figure>
    <figure
      class="image"
    >
      <div>
        <img
          alt="画像が読み込まれない場合はページを更新してみてください。"
          src="https://images.unsplash.com/photo-1512070679279-8988d32161be?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb"
        />
      </div>
      <figcaption
        class="caption"
      >
        External images from Unsplash
      </figcaption>
    </figure>
  </div>
</div>
`;
